### PR TITLE
feat(plugins): accumulated changes to documentation of the server application

### PR DESCRIPTION
**Pull request by str-anger**
Bringing server documentation and new naming model to the main branch.

### CODE REVIEW

**Comment by samiralavi** [comment]
@str-anger could you please update it with the main branch? It enables docs prereview for this PR.

File: docs/extensions.md [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+
+### What is an Extension Technically
+
+Technically, extensions are directories with files which follow an extension convention.
+This convention is described in the chapters below.
+
+When server executes an extension action, the following steps are followed:
+
+1. Server application (`aqueductcore` or `aqueductpro`) checks the correctness of the extension definition (see chapter on [Writing a Manifest File](#writing-a-manifest-file)).
+2. Then it check the existence of the python virtual environment inside extension's directory. If there is no virtual environment inside, it is created and populated with dependencies. You may control which modules are installed there by adding `requirements.txt` and/or `pyproject.toml` files. These dependencies are shared by all actions of the extension and installed once.
```
Comment (samiralavi) [inline docs/extensions.md:40]:
```suggestion
2. Then it checks the existence of the python virtual environment inside the extension's directory. If there is no virtual environment inside, it gets created and the extension's dependencies are installed. You may control which packages are installed by adding `requirements.txt` and/or `pyproject.toml` files to the extension's directory. These dependencies are shared by all actions of the extension and installed once.
```

File: docs/extensions.md [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+## What is Extension
+
+Extension is a piece of software, which adds custom functionality to the server instance 
+(`aqueductcore` or `aqueductpro`) without server update or restart.
+One extension may include multiple actions, connected by common settings or area of responsibility.
+For example, a set of graph plotting actions, integration with a 3rd-party service, or
+a collection of quantum circuit simulation methods.
+
+Extensions may be implemented in any programming language if this language interpreter 
+or binary architecture is supported by the executing (virtual) machine.
```
Comment (samiralavi) [inline docs/extensions.md:23]:
```suggestion
or binary architecture is supported by the executing (virtual) enviromnet.
```

File: docs/extensions.md [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+In the current implementation of extensions, their code is run on the Aqueduct server machine 
+(within server container, if you use containerised version). By default, `python3` and `bash` scripting  are supported.
```
Comment (samiralavi) [inline docs/extensions.md:25]: The container is fixed with Python 3.10, it is better to specify it.

File: docs/extensions.md [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+2. Then it check the existence of the python virtual environment inside extension's directory. If there is no virtual environment inside, it is created and populated with dependencies. You may control which modules are installed there by adding `requirements.txt` and/or `pyproject.toml` files. These dependencies are shared by all actions of the extension and installed once.
+3. Based on the manifest file and user input, parameters and constants are initialised as environment variables.
+4. `script` section of the action definition is executed. This may call arbitrary code: scripts or programs located in the directory. If `$python` variable is used inside this section, it is replaced with the python executable of the virtual environment. Read more about writing the code in [Writing Extension Code](#writing-extension-code) section. We expect that these programs read the data and persist the results inside the aqueduct experiment.
+5. Execution output (standard input stream, standard error stream, process result code) is recorded in a log file inside an experiment.
```
Comment (samiralavi) [inline docs/extensions.md:43]:
```suggestion
5. Execution output (standard input stream, standard error stream, process result code) is recorded in a log file which gets saved alongside the experiment's files.
```

File: docs/extensions.md [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+    export aqueduct_key=""
+    # TODO: add your parameters and constants below
+    export varname=123
+
+    # run
+    $venv/bin/python action.py
+    echo "Result code: $?"
+    ```
+
+Now, you may run and test your plugin action as `sh test.sh`. Do changes of variables and constants by adding them to the `test.sh` file, and reading them with `os.environ.get("name", "")` in python.
```
Comment (samiralavi) [inline docs/extensions.md:129]:
```suggestion
Now, you may run and test your extension action as `sh test.sh`. Do changes of variables and constants by adding them to the `test.sh` file, and reading them with `os.environ.get("name", "")` in python.
```

File: docs/extensions.md [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+        description: file which will be created inside experiment
+        data_type: str
+```
+
+Each manifest starts with a header with general extension information 
+(`name`, `description`, `authors`). This information is used to generate 
+the user interface. The `aqueduct_url` mandatory parameter is used to make 
+the extension aware of the instance of Aqueduct it should interact with. 
+In the current implementation, extensions are assumed to run on the same machine 
+as the server application, so this address will be a `http://localhost:8000/` or similar.
```
Comment (samiralavi) [inline docs/extensions.md:226]:
```suggestion
as the server application, so this address will be a `http://localhost:8000/` or similar. Port `8000` is the default port in the container that the Aqueduct server listens to.
```

File: docs/extensions.md [hunk 7]
```diff
@@ -0,0 +1,288 @@
...
+| `float`, `int`     | numerical types |
+| `bool`             | True or False. Passed to an extension action as 0 or 1 |
+
+Please, note, that purpose of extensions is to interact with an experiment. They may generate, process or consume the data from the experiment, that is why each extension action MUST have at least one parameter of `experiment` type.
+
+## Deploying an Extension
+
+Extensions are directories which could be distributed in the form of archives. When starting the server using Docker, 
+bind mount a host directory to the Aqueduct container at `/workspace/extensions`.
+Unpack an extension directory into the host directory, and the container will immediately see the changes.
```
Comment (samiralavi) [inline docs/extensions.md:275]: We should add that this directory should be provided as an environment variable. `/workspace/extensions`. Also, it says in a way that `/workspace/extensions` is a hardcode path that Aqueduct expects, which is not.

**Review by samiralavi — APPROVED** [review]
Great work @str-anger, it is a complete reference with good tutorial points. I have put some suggestions to improve it and avoid confusion for the readers.

File: docs/extensions.md [hunk 8]
```diff
@@ -0,0 +1,288 @@
...
+summary: this document describes extension programming, installation, and setup process.
+---
+
+On this page we provide information related to extension development, deployment and setup:
+
+- [What is Extension](#what-is-extension)
+- [Writing Extension Code](#writing-extension-code)
+- [Writing a Manifest File](#writing-a-manifest-file)
+- [Deploying an Extension](#deploying-an-extension)
+- [Extension Setup](#extension-setup)
```
Comment (safeamiiir) [inline docs/extensions.md:12]: Maybe a minor change in the section title?
```suggestion
- [What is an Extension](#what-is-extension)
- [Writing an Extension](#writing-extension-code)
- [Writing a Manifest File](#writing-a-manifest-file)
- [Deploying an Extension](#deploying-an-extension)
- [Extension Setup](#extension-setup)
```

**Review by safeamiiir — COMMENTED** [review]
Thanks Stas! And It has been already reviewed in other PRs.

I was just wondering we might merge it after other changes are being merged, because this PR includes some unrelated changes as `/docs/cli`

**Pull request merged**